Use express instead of native http module (#485)
discordjs/guide: guide/oauth2/README.md

###  Setting up a basic web server

Most of the time, OAuth2 is used in websites to get information about its users from an external service. In this example, you will use `express` module to create a web server to use a user's Discord information to greet them. Firstly, you should install `express`.

```bash
$ npm install express
```

And then, create a file named `index.js` which will be used to start the server.

```js
const express = require('express');
const app = express();
const fs = require('fs');
const port = 53134;

app.get('/', (request, response) => {

> 	response.status(200).send('index.html', { root: '.' });

Reviewer: ```suggestion
	response.status(200).sendFile('index.html', { root: '.' });
```

});

> // 404 Route, Keep this route as the last routes

Reviewer: ```suggestion
// Catch-all route; keep this as the last route
```

	<a id="login" style="display: none;" href="your oauth2 url here">Identify Yourself</a>
	<script>
		window.onload = () => {
			const fragment = parseQuery(window.location.href);

			if (fragment.access_token) {
				const accessToken = fragment.access_token;
				const tokenType = fragment.token_type;

> 				fetch('https://discordapp.com/api/users/@me', {

Reviewer 2: ```suggestion
				fetch('https://discord.com/api/users/@me', {
```

				document.getElementById('login').style.display = 'block';
			}
		}
		function parseQuery(queryString) {
            var query = {};
            var pairs = (queryString[0] === '?' ? queryString.substr(1) : queryString).split('&');
            for (var i = 0; i < pairs.length; i++) {
                var pair = pairs[i].split('=');
                query[decodeURIComponent(pair[0])] = decodeURIComponent(pair[1] || '');
            }
            return query;

>         }

Reviewer 3: same as previous suggestion in the other example

#### Authorization code grant

> Unlike the quick example, you need an OAuth2 url where the `response_type` is `code`. Once you've obtained it, try visiting the link and authorizing your application. You should notice that instead of a hash, the redirect url now has a single query parameter appended to it like `?code=ACCESS_CODE`. Modify your `index.js` (on `/` route) file to pull the parameter out of the url if it exists.

Reviewer: ```suggestion
Unlike the quick example, you need an OAuth2 url where the `response_type` is `code`. Once you've obtained it, try visiting the link and authorizing your application. You should notice that instead of a hash, the redirect url now has a single query parameter appended to it like `?code=ACCESS_CODE`. Modify your `index.js` (on the `/` route) file to pull the parameter out of the url if it exists.
```

Make the route as a async function:
```js
app.get('/', async (request, response) => {
	// ...
});
```

Then:
<!-- eslint-skip -->
```js
let access_token;
let token_type;

if (request.query.code) {
	const accessCode = request.query.code;
	console.log(`The access code is: ${accessCode}`);
}

if (token_type && access_token && request.query.state) {

> 	response.status(200).redirect(`/?code=${request.query.code}&state=${request.query.state}&access_token=${access_token}&token_type=${token_type}`);

Reviewer: This if statement should be moved inside the `if (request.query.code)` statement, because it utilizes the code part of the URI, as well as other data from the `await res.json()` call. You also use `request.query.code` here instead of the `accessCode` variable created above.

Reviewer: Actually, I don't think this redirect is needed. You need the access token/token type/state query params in the index.html file, but only when using the the `response_type=token` param used in the index.html code at first. When moving it over to the server side, it shows how to implement the OAuth code and switches the `response_type` to `code`.  At the bottom of the page it also mentions:
> To maintain security, store the access token server side but associate it with a session ID that you generate for the user.

> } else { response.sendFile('index.html', { root: '.' }); }

Reviewer: Either make this an early return or place the `else {}` content on its own line.

```

> Now you have to exchange this code with Discord for an access token. To do this, you need your `client_id` and `client_secret`. If you've forgotten them, head over to [your applications](https://discordapp.com/developers/applications) and get them. You can use `node-fetch` to make requests to Discord; you can install it with `npm i node-fetch`. Place this code after `request.query.code` statement.

Reviewer: ```suggestion
Now you have to exchange this code with Discord for an access token. To do this, you need your `client_id` and `client_secret`. If you've forgotten them, head over to [your applications](https://discord.com/developers/applications) and get them. You can use `node-fetch` to make requests to Discord; you can install it with `npm i node-fetch`. Place this code after your `if (request.query.code)` statement.
```

	scope: 'the scopes',
};

> const res = await fetch('https://discordapp.com/api/oauth2/token', {

Reviewer 2: ```suggestion
const res = await fetch('https://discord.com/api/oauth2/token', {
```

	method: 'POST',
	body: new URLSearchParams(data),
	headers: {
		'Content-Type': 'application/x-www-form-urlencoded',
	},
});

> const info = await res.json();

Reviewer: ```suggestion
const { token_type: tokenType, access_token: accessToken } = await res.json();
```

Update the relevant code as well.

```

Now that you have an access token and a refresh token, try fetching the user's information. It's the exact same as how it was done in the html file.
<!-- eslint-skip -->
```js

> const userPost = await fetch('https://discordapp.com/api/users/@me', {

Reviewer 2: ```suggestion
const userPost = await fetch('https://discord.com/api/users/@me', {
```
